modules/dev-utils/src/main/java/org/apache/ignite/development/utils/IgniteIndexReader.java

```diff
@@ -0,0 +1,2130 @@
+
+                            listIds.forEach(listId -> allPages.addAll(getPageList(listId, pageListStat)));
+
+                            bucketsData.put(new IgniteBiTuple<>(fNextMetaId, k), listIds);
+                        });
+
+                        nextMetaId = io.getNextMetaPageId(addr);
+                    }
+                    catch (Exception e) {
+                        errors.put(nextMetaId, singletonList(e));
+
+                        nextMetaId = 0;
+                    }
+                }
+
+                return null;
+            });
+        }
+        catch (IgniteCheckedException e) {
+            throw new IgniteException(e);
+        }
+
+        return new PageListsInfo(bucketsData, allPages, pageListStat, errors);
+    }
+
+    /**
+     * Get single page list.
+     *
+     * @param pageListStartId Id of the start page of the page list.
+     * @param pageStat Page types statistics.
+     * @return List of page ids.
+     */
+    private List<Long> getPageList(long pageListStartId, Map<Class, AtomicLong> pageStat) {
+        List<Long> res = new LinkedList<>();
+
+        long nextNodeId = pageListStartId;
+
+        ByteBuffer nodeBuf = allocateBuffer(pageSize);
+        ByteBuffer pageBuf = allocateBuffer(pageSize);
+
+        long nodeAddr = bufferAddress(nodeBuf);
+        long pageAddr = bufferAddress(pageBuf);
+
+        try {
+            while (nextNodeId != 0) {
+                try {
+                    nodeBuf.rewind();
+
+                    readPage(idxStore, nextNodeId, nodeBuf);
+
+                    PagesListNodeIO io = PageIO.getPageIO(nodeAddr);
+
+                    for (int i = 0; i < io.getCount(nodeAddr); i++) {
+                        pageBuf.rewind();
+
+                        long pageId = normalizePageId(io.getAt(nodeAddr, i));
+
+                        res.add(pageId);
+
+                        readPage(idxStore, pageId, pageBuf);
+
+                        PageIO pageIO = PageIO.getPageIO(pageAddr);
+
+                        pageStat.computeIfAbsent(pageIO.getClass(), k -> new AtomicLong(0)).incrementAndGet();
+                    }
+
+                    nextNodeId = io.getNextId(nodeAddr);
+                }
+                catch (IgniteCheckedException e) {
+                    throw new IgniteException(e.getMessage(), e);
+                }
+            }
+        }
+        finally {
+            freeBuffer(nodeBuf);
+            freeBuffer(pageBuf);
+        }
+
+        return res;
+    }
+
+    /**
+     * Traverse all trees in file and return their info.
+     *
+     * @param metaTreeRoot Meta tree root page id.
+     * @return Index trees info.
+     */
+    private Map<String, TreeTraversalInfo> traverseAllTrees(
+        String traverseProcCaption,
+        long metaTreeRoot,
+        Supplier<ItemStorage> itemStorageFactory,
+        TraverseProc traverseProc
+    ) {
+        Map<String, TreeTraversalInfo> treeInfos = new HashMap<>();
```

Comment:
Better to use LinkedHashMap here, because I want to see MetaTree at head of view.